From 280792bfdc51ec199482abc28b1cd4b0e4ee4988 Mon Sep 17 00:00:00 2001
From: Frederic Danis <frederic.danis@intel.com>
Date: Mon, 14 Apr 2014 12:11:02 +0200
Subject: [PATCH] net: fake socket ioctl to reset connections matching local
 address

Android adds a new socket ioctl SIOCKILLADDR , but current implementation
(commit 567620f5968380fc43231ef9806c5bee59eb8d2d) was reverted has it
stall CPU (commit af1f47d18fbb4db67b4570b9bd8f73f0529392a7).

This adds a fake implementation for this ioctl.

Change-Id: I488444c61a0284b2525267030bfbb2826cb1a887
Signed-off-by: Frederic Danis <frederic.danis@intel.com>
---
 include/uapi/linux/sockios.h | 1 +
 net/ipv4/af_inet.c           | 1 +
 net/ipv4/devinet.c           | 7 ++++++-
 net/ipv6/af_inet6.c          | 4 ++++
 4 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/include/uapi/linux/sockios.h b/include/uapi/linux/sockios.h
index e888b1a..623e9aa 100644
--- a/include/uapi/linux/sockios.h
+++ b/include/uapi/linux/sockios.h
@@ -65,6 +65,7 @@
 #define SIOCDIFADDR	0x8936		/* delete PA address		*/
 #define	SIOCSIFHWBROADCAST	0x8937	/* set hardware broadcast addr	*/
 #define SIOCGIFCOUNT	0x8938		/* get number of devices */
+#define SIOCKILLADDR	0x8939		/* kill sockets with this local addr */
 
 #define SIOCGIFBR	0x8940		/* Bridging support		*/
 #define SIOCSIFBR	0x8941		/* Set bridging options 	*/
diff --git a/net/ipv4/af_inet.c b/net/ipv4/af_inet.c
index 7b7a25be8..26fe9f4 100644
--- a/net/ipv4/af_inet.c
+++ b/net/ipv4/af_inet.c
@@ -889,6 +889,7 @@ int inet_ioctl(struct socket *sock, unsigned int cmd, unsigned long arg)
 	case SIOCSIFPFLAGS:
 	case SIOCGIFPFLAGS:
 	case SIOCSIFFLAGS:
+	case SIOCKILLADDR:
 		err = devinet_ioctl(net, cmd, (void __user *)arg);
 		break;
 	default:
diff --git a/net/ipv4/devinet.c b/net/ipv4/devinet.c
index bdbf68b..c0167be 100644
--- a/net/ipv4/devinet.c
+++ b/net/ipv4/devinet.c
@@ -929,6 +929,7 @@ int devinet_ioctl(struct net *net, unsigned int cmd, void __user *arg)
 	case SIOCSIFBRDADDR:	/* Set the broadcast address */
 	case SIOCSIFDSTADDR:	/* Set the destination address */
 	case SIOCSIFNETMASK: 	/* Set the netmask for the interface */
+	case SIOCKILLADDR:	/* Kill sockets with this local address */
 		ret = -EPERM;
 		if (!ns_capable(net->user_ns, CAP_NET_ADMIN))
 			goto out;
@@ -980,7 +981,8 @@ int devinet_ioctl(struct net *net, unsigned int cmd, void __user *arg)
 	}
 
 	ret = -EADDRNOTAVAIL;
-	if (!ifa && cmd != SIOCSIFADDR && cmd != SIOCSIFFLAGS)
+	if (!ifa && cmd != SIOCSIFADDR && cmd != SIOCSIFFLAGS
+	    && cmd != SIOCKILLADDR)
 		goto done;
 
 	switch (cmd) {
@@ -1107,6 +1109,9 @@ int devinet_ioctl(struct net *net, unsigned int cmd, void __user *arg)
 			inet_insert_ifa(ifa);
 		}
 		break;
+	case SIOCKILLADDR:
+		ret = 0;
+		break;
 	}
 done:
 	rtnl_unlock();
diff --git a/net/ipv6/af_inet6.c b/net/ipv6/af_inet6.c
index d753676..fa9e4e4 100644
--- a/net/ipv6/af_inet6.c
+++ b/net/ipv6/af_inet6.c
@@ -513,6 +513,10 @@ int inet6_ioctl(struct socket *sock, unsigned int cmd, unsigned long arg)
 		return addrconf_del_ifaddr(net, (void __user *) arg);
 	case SIOCSIFDSTADDR:
 		return addrconf_set_dstaddr(net, (void __user *) arg);
+	case SIOCKILLADDR:
+		if (!capable(CAP_NET_ADMIN))
+			return -EACCES;
+		return 0;
 	default:
 		if (!sk->sk_prot->ioctl)
 			return -ENOIOCTLCMD;
-- 
1.8.3.2

